Created RoundImageEx, Moved ImageEx Core Logic to Base Class

WilliamABradley opened the pull request:
I was looking for a way to clip an `ImageEx` Control, then I stumbled on to #277 where @JohnnyWestlake suggested restructuring the `ImageEx` to use `ImageBrush`es, but didn't want to because it would remove features.

He wanted to have the Original ImageEx Logic and I wanted the ability to make a Circular Image. 
I thought, Why don't we have both?
Da, Da Da, Da Da, Da.
![alt Meme](http://i0.kym-cdn.com/photos/images/newsfeed/000/538/731/0fc.gif)

I shifted all of the Core Logic to the `ImageExBase` Base Class, `ImageEx` retains complete original functionality, and I created the new `RoundImageEx` to use `Ellipse`s and `ImageBrush`es.

`RoundImageEx` loses `GetAsCastingSource()` & `NineGrid`, however, it gains the ability to add a `Stroke` w/ `StrokeThickness` & the ability to disable Stroke on Placeholder Images. 

I also updated the ImageEx Sample Page, although it doesn't demonstrate using Stroke, or other `RoundImageEx` properties.

dnfclas commented:
@__WilliamABradley__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

dnfclas commented:
__@WilliamABradley__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

WilliamABradley commented:
It all works except for the View State to show StrokeThickness on the Placeholder Image. Currently I have just set the StrokeThickness to the Placeholder Image, so the ShowPlaceholderStroke bool does nothing.

hermitdave commented:
@WilliamABradley I think something has wrong very wrong. Another PR is being referenced. That way you need to do is that the 2nd PR must start on dev again

WilliamABradley commented:
@hermitdave How do you mean? This one was on the dev branch, but I wanted to create a second PR, but it wouldn't let me take a branch off of Microsoft/Dev, so I branched off of Origin/Dev and reverted the changes. How do I fix this mess?

Do you mean #1109? as This PR doesn't have any references to this PR, but that PR has references to this PR. Help :( I'm a GitHub noob.

hermitdave commented:
maybe I didn't pay enough attention..

it however looks to me that there are conflicts  for 
Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageEx.Source.cs 
Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageEx.cs

WilliamABradley commented:
@hermitdave I shifted the Core Logic out of ImageEx into a base class so that a RoundImageEx class can use the shared Logic. (The ImageBrush and Image implementations are different, so an XAML restyle wouldn't work here). ImageEx gets to retain all of it's functionality, and RoundImageEx gets to use ImageCache and most of the features of ImageEx, except it loses Nine Grid and GetAsCasting source, but you can use it to create circular images, which ImageEx can't do.

WilliamABradley commented:
@hermitdave I Merged back onto Microsoft/Dev if that helps?
Perhaps a better name for this control is ImageBrushEx?

hermitdave commented:
@WilliamABradley you need to run Please run '.\build.ps1 UpdateHeaders' from powershell to fix issue with file headers. I have reviewed much of the code yesterday. Will go through again over the weekend

WilliamABradley commented:
[![psError.png](https://s20.postimg.org/oh2lzkszx/ps_Error.png)](https://postimg.org/image/4z7yjmw21/)

@hermitdave  I'm getting this issue. Is it compatible with VS 2017? Or do I need something else?

Odonno commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.Placeholder.cs:
```diff
@@ -52,6 +50,29 @@ public ImageSource PlaceholderSource
+
+            if (e.OldValue == null || e.NewValue == null || !e.OldValue.Equals(e.NewValue))
+            {
+                if (control is RoundImageEx round)
```
From what I know, C# 7 is not currently available for Continuous Integration. Please change to use C# 6 features such as `as`.

WilliamABradley replied:
:( I looked it up and I couldn't find anything about it? It is located in quite a bit of my code, in this and in #1116, so it would be a pain to replace.

Odonno replied:
@WilliamABradley If you want us to merge your branch, I recommend you to retarget your code to C# 6 only. To clarify, I am not against C# 7 but the fact is that the solution will not compile because CI does not support C# 7 for the moment.

hermitdave replied:
think you might be better off with something like this
```
RoundImageEx round = control as RoundImageEx
if (round != null)
{
}
else
{
}
```

WilliamABradley replied:
Fixed it, I'll do the same for my other PR.

WilliamABradley commented:
@Odonno @hermitdave I have replaced the C# 7 features with C# 6 compatible features. I have also fixed an issue that caused a crash in .Net Native. I believe that this PR is ready for final review, as it is also now passing in AppVeyor :)

WilliamABradley commented:
@Odonno What happens now?

hermitdave commented:
A PR requires 2 reviews, I'll do mine maybe tonight. If all is well, i will merge it in

WilliamABradley commented:
@hermitdave Awesome! If when I get around to it, I also have a few more things I could contribute, such as:
- A Helper for ContentDialogs to ensure only one is open at any time, a way to interrupt a dialog with another, if it is more important, and a way to interrupt a dialog and afterwards return to the previous dialog.
- Another Helper for Producing a Cortana VCD.xml via Code (Currently CortanaExtensions on GitHub), based on the idea of NotificationExtensions, it allows you to easily populate the VCD at runtime, as well as populate it with translated strings.
- A Control similar to AdaptiveGridView, except that it maintains aspect ratio, and uses ViewBoxes to scale the Items.

hermitdave commented:
@WilliamABradley create issues and uservoice entries for those.

As far as adaptive grid like control, maybe providing that as an option within AdaptiveGridView might be desirable.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageEx.Members.cs:
```diff
@@ -10,14 +10,12 @@
 {
+    using Windows.Media.Casting;
+    using Windows.UI.Xaml;
+    using Windows.UI.Xaml.Controls;
```
usings should be outside of the namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.Members.cs:
```diff
@@ -0,0 +1,108 @@
+namespace Microsoft.Toolkit.Uwp.UI.Controls
+{
+    using System;
+    using Windows.UI.Xaml;
```
usings outside of namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.Placeholder.cs:
```diff
@@ -10,34 +10,32 @@
-
 namespace Microsoft.Toolkit.Uwp.UI.Controls
 {
+    using Windows.UI.Xaml;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.Placeholder.cs:
```diff
@@ -52,6 +50,30 @@ public ImageSource PlaceholderSource
+
+            if (e.OldValue == null || e.NewValue == null || !e.OldValue.Equals(e.NewValue))
+            {
+                RoundImageEx round = control as RoundImageEx;
```
Discussion: There should be a virtual OnPlaceholderSourceChanged method that RoundImageEx can override and all this code should go in there, rather than be in the base class.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.Source.cs:
```diff
@@ -10,28 +10,26 @@
 {
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExBase.cs:
```diff
@@ -0,0 +1,164 @@
+namespace Microsoft.Toolkit.Uwp.UI.Controls
+{
+    using System;
+    using Windows.Foundation;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExFailedEventArgs.cs:
```diff
@@ -10,10 +10,10 @@
-
 namespace Microsoft.Toolkit.Uwp.UI.Controls
 {
+    using System;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/ImageExOpenedEventArgs.cs:
```diff
@@ -10,10 +10,10 @@
-
 namespace Microsoft.Toolkit.Uwp.UI.Controls
 {
+    using System;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.Members.cs:
```diff
@@ -0,0 +1,64 @@
+namespace Microsoft.Toolkit.Uwp.UI.Controls
+{
+    using Windows.UI.Xaml;
+    using Windows.UI.Xaml.Media;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.cs:
```diff
@@ -0,0 +1,44 @@
+namespace Microsoft.Toolkit.Uwp.UI.Controls
+{
+    using Windows.UI.Xaml;
+    using Windows.UI.Xaml.Media;
```
usings outside namespace

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.xaml:
```diff
@@ -0,0 +1,125 @@
+﻿<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
```
Ensure xaml styler has been run on this.

WilliamABradley commented:
@ScottIsAFool Those changes have been made, however, `ShowPlaceholderStroke` currently does nothing, because I couldn't get it to work. The `StrokeThickness` on `PlaceholderEllipse` should default to 0, With the Visual State `ShowStroke` setting it to the desired Stroke value. Any ideas as to why it doesn't work?

ScottIsAFool commented:
@WilliamABradley you can't have a binding in the To value when used inside a control template. This would have to be set in code, sadly.

WilliamABradley commented:
@ScottIsAFool How would that be done? Would you attach an x:Name to the Animation so you update the value when the StrokeThickness changes, or some other way?

ScottIsAFool commented:
@WilliamABradley to be honest, looking at it, I'm not sure any of that is needed. Your ellipse's StrokeThickness is already being set and that should suffice. I don't think the ShowPlaceholderStroke property is even needed. I'm going to add a couple more comments to my review.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.xaml:
```diff
@@ -0,0 +1,125 @@
+                                 HorizontalAlignment="{TemplateBinding HorizontalAlignment}"
+                                 VerticalAlignment="{TemplateBinding VerticalAlignment}"
+                                 Opacity="1.0"
+                                 Stroke="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=Stroke}"
```
This should just be a TemplateBinding

WilliamABradley replied:
This is due to a bug in ImageBrushes http://stackoverflow.com/questions/22134281/templatebinding-fails-when-the-target-is-an-imagebrush-imagesource

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.xaml:
```diff
@@ -0,0 +1,125 @@
+                                 VerticalAlignment="{TemplateBinding VerticalAlignment}"
+                                 Opacity="1.0"
+                                 Stroke="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=Stroke}"
+                                 StrokeThickness="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=StrokeThickness}">
```
This should also just be a TemplateBinding

WilliamABradley replied:
This is due to a bug in ImageBrushes http://stackoverflow.com/questions/22134281/templatebinding-fails-when-the-target-is-an-imagebrush-imagesource

ScottIsAFool replied:
I'm not sure this is applicable. From what that SO says, this is because the OP needed to basically convert to another value. As long as Stroke and StrokeThickness on RoundImageEx are using the same property types as Ellipse, then no conversion should be needed and TemplateBinding can be used.

WilliamABradley replied:
Woops, It was only for the ImageBrushes anyway, not the Properties on the Ellipse.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.xaml:
```diff
@@ -0,0 +1,125 @@
+                                </VisualState>
+                                <VisualState x:Name="Unloaded" />
+                            </VisualStateGroup>
+                            <VisualStateGroup x:Name="StrokeStates">
```
I'm not sure this VisualState is even needed.

WilliamABradley replied:
Removed.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.Members.cs:
```diff
@@ -0,0 +1,64 @@
+            DependencyProperty.Register(nameof(StrokeProperty), typeof(Brush), typeof(RoundImageEx), new PropertyMetadata(default(Brush)));
+
+        // Using a DependencyProperty as the backing store for ShowPlaceholderStroke.  This enables animation, styling, binding, etc...
+        public static readonly DependencyProperty ShowPlaceholderStrokeProperty =
```
Is this property even needed? If the dev doesn't want the placeholder to have a stroke, then the StrokeThickness should just be set to 0.

WilliamABradley replied:
Makes sense

ScottIsAFool commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.Members.cs:
```diff
@@ -0,0 +1,64 @@
+    {
+        // Using a DependencyProperty as the backing store for StrokeThickness.  This enables animation, styling, binding, etc...
+        public static readonly DependencyProperty StrokeThicknessProperty =
+            DependencyProperty.Register(nameof(StrokeThickness), typeof(double), typeof(RoundImageEx), new PropertyMetadata(default(double)));
```
This property should really be using Thickness, not double as its type.

WilliamABradley replied:
The StrokeThickness on an Ellipse is a double Property, this is just a pass-through. It wouldn't make sense for a circle to have differing thickness on different sides. It has no sides,

ScottIsAFool replied:
Interesting, ignore this. Apparently for Ellipse StrokeThickness uses double, not Thickness, like everywhere else does 8-)

hermitdave commented on Microsoft.Toolkit.Uwp.UI.Controls/ImageEx/RoundImageEx.xaml:
```diff
@@ -0,0 +1,111 @@
+            <Setter.Value>
+                <ControlTemplate TargetType="controls:RoundImageEx">
+                    <Grid>
+                        <Ellipse x:Name="PlaceholderEllipse"
```
I know its a bit late in the day. Any reason we are using Ellipse instead of Border ?  Border would allow you to have image with rounded corner rather than elliptical images which is what default will be unless the user were to go override the XAML.

Personally Image with CornerRadius better than PeoplePicture (the one coming out in FCU)

WilliamABradley replied:
I guess that could work too, because you could set the corner radius to say 999, in order to create a round image anyway right?

hermitdave replied:
Its not difficult to create rounded image by tweaking corner radius. I am not sure if there's a default value that forces circular ellipse but my point is Border with RoundedCorner can be turned into fully circular image if desired. 
Ellipse can't do that. Any thoughts on why we should go for one or the other ?

WilliamABradley replied:
The Control name wouldn't make sense anymore. I could name it ImageBrushEx, although it would still be encapsulated so it wouldn't really be an ImageBrush you could attach to things,

hermitdave replied:
good point about naming. @nmetulev @ScottIsAFool @Odonno any thoughts on this ?

WilliamABradley replied:
Another thing is, without setting the corner radius, it behaves practically the same as ImageEx, but without `NineGrid` and `GetAsCastingSource`

WilliamABradley commented:
@nmetulev I have updated the documentation.

nmetulev commented:
Thanks :)
